### PR TITLE
Fix for mqtt reconnection problems

### PR DESCRIPTION
**Pull request by MyIgel**
After some testing and guessing this seems to fix the reconnect problems.

## Problem
* The client connects to the broker
* Forgets this connection / drops it <- I'm not sure why but this breaks the state machine
* Reconnects to the broker
* The broker closes the old connection and notifies the client
* The client drops his new connection & reconnects

The whole cycle takes about 3 seconds, sometimes the client is able to send and receive some data, sometimes not.

## Server Log
```
mqtt mosquitto[8372]: Client asdfasdfasdf already connected, closing old connection.
```

And a pcap of the two connections: [mqtt-server.pcap.zip](https://github.com/esper-hub/esper/files/2172858/mqtt-broken.zip)

## Client Log
```
[...]
[mqtt] Publishing message to esper/dfasdf/stripe/brightness/get : 169
[mqtt] Publishing message to esper/dfasdf/stripe/rgb/get : 132,132,132
[mqtt] Publishing message to esper/dfasdf/stripe/effect/get : rainbowFade
[mqtt] Received topic: esper/dfasdf/stripe/rgb/set
[mqtt] Received message: 132,132,132
[mqtt] Publishing message to esper/dfasdf/stripe/brightness/get : 169
[mqtt] Publishing message to esper/dfasdf/stripe/rgb/get : 132,132,132
[mqtt] Publishing message to esper/dfasdf/stripe/effect/get :
[mqtt] Connecting 
[device] MQTT state changed: Connecting
[mqtt] Disconnected
[device] MQTT state changed: Disconnected
[heartbeat] Stop awaiting heartbeats
[device] MQTT state changed: Connected \o/
[mqtt] Publishing message to esper/dfasdf/status : ONLINE
[mqtt] Subscribed to: esper/heartbeat
[mqtt] Subscribed to: esper/update
[mqtt] Subscribed to: esper/dfasdf/update
[mqtt] Subscribed to: esper/reboot
[mqtt] Subscribed to: esper/dfasdf/relay1/set
[mqtt] Subscribed to: esper/dfasdf/relay2/set
[mqtt] Subscribed to: esper/dfasdf/light/set
[mqtt] Subscribed to: esper/dfasdf/stripe/brightness/set
[mqtt] Subscribed to: esper/dfasdf/stripe/effect/set
[mqtt] Subscribed to: esper/dfasdf/stripe/rgb/set
[mqtt] Subscribed to: esper/dfasdf/stripe/command/send
[heartbeat] Start awaiting heartbeats
[info] Publishing device info
[mqtt] Publishing message to esper/dfasdf/info : {
[...]
```

### CODE REVIEW

File: framework/managers/MqttConnectionManager.cpp [hunk 1]
```diff
@@ -61,8 +61,12 @@ void MqttConnectionManager::onDisconnected(TcpClient &client, bool flag) {
         LOG.log("Unreachable");
     }
 
-    this->state.set(State::DISCONNECTED);
-    this->reconnectTimer.start();
+    if (this->getState() != State::CONNECTING) {
```
Comment (fooker) [inline framework/managers/MqttConnectionManager.cpp]: You can use `this->state` here as it casts to the wrapped value

Comment (fooker) [inline framework/managers/MqttConnectionManager.cpp]: We should ignore disconnects while being disconnected, too.

File: framework/managers/MqttConnectionManager.cpp [hunk 2]
```diff
@@ -61,8 +61,12 @@ void MqttConnectionManager::onDisconnected(TcpClient &client, bool flag) {
...
+        this->state.set(State::DISCONNECTED);
+        this->reconnectTimer.start();
+    } else {
+        LOG.log("Ignored disconnect while connecting");
```
Comment (fooker) [inline framework/managers/MqttConnectionManager.cpp]: As this is a workaround we should use a log message which is more alerting. At least prefix with `Error: ` or something like that.

**Comment by MyIgel** [comment]
The problem is that this happens random, i'v seen a device reconnecting after a minute and also some after days, so sitting and staring at logs is not the best idea ;)

I have done that with a pi but that needs a resetup (and i currently have no sonoff laying around)

File: framework/managers/MqttConnectionManager.cpp (hunk 1 above)
Comment (MyIgel) [reply framework/managers/MqttConnectionManager.cpp]: Hm, while being disconnected that would only restart the timer? So it shouldn't be a problem but ok

Comment (MyIgel) [reply framework/managers/MqttConnectionManager.cpp]: Totally missed that... :D

Comment (MyIgel) [reply framework/managers/MqttConnectionManager.cpp]: And should I check for flag too? If the broker is not reachable it should be false and the state should be CONNECTING, hmm

Comment (fooker) [reply framework/managers/MqttConnectionManager.cpp]: I think checking for `== CONECTED` should be fine.

**Comment by fooker** [comment]
We have seen this today in our environment, too.

A restart of `mosquitto` seems to fix this for some time. The interesting thing is, that other MQTT clients seem to be affected, too. The do not suffer from the duplicated client problem but get disconnected / reconnected by mosquitto.

I will merge the PR after some testing to circumvent the problem, but it seems like `mosquitto` must be fixed, too.

**Comment by fooker** [comment]
Merged.

**Pull request closed without merging**